src/inspect_ai/tool/_tools/_computer/_computer.py

```diff
@@ -0,0 +1,200 @@
+          The output of the command. Many commands will include a screenshot reflecting the result of the command in their response.
+        """
+        parts = cmd.split(maxsplit=1)
+        if not parts:
+            raise ValueError("Empty command")
+
+        action = parts[0]
+        arguments = parts[1] if len(parts) > 1 else ""
+
+        async def key_action(args: str) -> ToolResult:
+            return await press_key(args)
+
+        async def type_action(args: str) -> ToolResult:
+            return await type(validate_type(args))
+
+        async def cursor_position_action(args: str) -> ToolResult:
+            validate_no_args(args, "cursor_position")
+            return await cursor_position()
+
+        async def mouse_move_action(args: str) -> ToolResult:
+            x, y = validate_coordinates(args, "mouse_move")
+            return await mouse_move(x, y)
+
+        async def left_click_action(args: str) -> ToolResult:
+            validate_no_args(args, "left_click")
+            return await left_click()
+
+        async def left_click_drag_action(args: str) -> ToolResult:
+            endx, endy = validate_coordinates(args, "left_click_drag")
+            return await left_click_drag(endx, endy)
+
+        async def right_click_action(args: str) -> ToolResult:
+            validate_no_args(args, "right_click")
+            return await right_click()
+
+        async def middle_click_action(args: str) -> ToolResult:
+            validate_no_args(args, "middle_click")
+            return await middle_click()
+
+        async def double_click_action(args: str) -> ToolResult:
+            validate_no_args(args, "double_click")
+            return await double_click()
+
+        async def screenshot_action(args: str) -> ToolResult:
+            validate_no_args(args, "screenshot")
+            return await screenshot()
+
+        # Dictionary to map actions to functions
+        action_map: dict[str, ActionFunction] = {
+            "key": key_action,
+            "type": type_action,
+            "cursor_position": cursor_position_action,
+            "mouse_move": mouse_move_action,
+            "left_click": left_click_action,
+            "left_click_drag": left_click_drag_action,
+            "right_click": right_click_action,
+            "middle_click": middle_click_action,
+            "double_click": double_click_action,
+            "screenshot": screenshot_action,
+        }
+
+        # Execute the corresponding function
+        try:
+            if action in action_map:
+                result: ToolResult | Awaitable[ToolResult] = action_map[action](
+                    arguments
+                )
+                return (await result) if asyncio.iscoroutine(result) else result  # type: ignore
+            else:
+                available_actions = ", ".join(sorted(action_map.keys()))
+                return f"Unknown action: {action}. Available actions are: {available_actions}"
+        except Exception as e:
```

Comment:
Here we should raise a `ToolError` (which will end up delivering the same message to the model but with an `error` bit set that might be meaningful/useful).

Ideally, we could distinguish between errors that could occur in the normal course of computer (which we report to the model with `ToolError`) and truly unexpected errors that should hard fail the eval. Not sure though whether we have sufficient context to do this though. FWIW the web browser always returns a `ToolError` unless there is a protocol violation (e.g. no msg or error returned from RPC).

Reply:
I've updated the code, but let's walk through it (including `_send_cmd` in the lower layer) together in person.